DBZ-4401 Support partition key row deletion and range tombstone row types

smiklosovic opened the pull request:
Hi @gunnarmorling , @bingqinzhou , @ahmedjami 

This patch adds support for partition key row deletion and range tombstones.

For range tombstones, it is introducing new struct called "bounds" with "start" and "end" fields.

This is the implementation for Cassandra 4 only for now. If we agree with the approach, I will try to introduce this to Cassandra 3 too.

There are added tests for each scenario for Cassandra 4 implementation.

Please tell me what you think about this.

smiklosovic commented on cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogReadHandlerImpl.java:
```diff
@@ -288,8 +295,17 @@ private void process(PartitionUpdate pu, OffsetPosition offsetPosition, Keyspace
                         LOGGER.warn("Encountered an unsupported row type {}, skipping...", rowType);
                         continue;
                     }
-                    Row row = (Row) rowOrRangeTombstone;
```
Previous code had it wrong. For Cassandra 4, we can not cast `RangeTombstoneBoundMarker` to `Row`. In Cassandra 4, `Row` implements `Unfiltered` and `RangeTombstoneBoundMarker` implements `Unfiltered`, but `RangeTombstoneBoundMarker` does not extend nor implement `Row` so we do not have any information about values but the primary key in `PartitionUpdate`.

ahmedjami commented:
Hi,
Thanks a lot @smiklosovic for this new PR :) 
I'm back and I'm planning to test the new feature soon.
Cheers,
Ahmed

smiklosovic commented:
hey @ahmedjami wait a day or two and I ll update this stuff on top of the current main branch, this one is little bit behind.

smiklosovic commented:
@ahmedjami  it is aligned, you can try now.

jpechane commented:
@vjuranek Could you please take a look?
@ahmedjami I'd be happy to hear about your feedback.

ahmedjami commented:
Hey!

> @ahmedjami I'd be happy to hear about your feedback.
Yes ;)

vjuranek commented on cassandra-4/src/test/java/io/debezium/connector/cassandra/RangeTombstoneCommitLogProcessingTest.java:
```diff
@@ -0,0 +1,70 @@
+
+        Record range1 = (Record) events.get(2);
+        assertEquals(range1.getEventType(), CHANGE_EVENT);
+        assertEquals(RANGE_TOMBSTONE, range1.getOp());
```
would be good to test also `start` and `end` of the range delete

vjuranek commented on cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogReadHandlerImpl.java:
```diff
@@ -416,7 +520,7 @@ private void populatePartitionColumns(RowData after, PartitionUpdate pu) {
                 String name = columns.get(i).name.toString();
                 Object value = partitionKeys.get(columnMetadata.position());
-                CellData cellData = new CellData(name, value, null, CellData.ColumnType.PARTITION);
+                CellData cellData = new CellData(name, value, null, PARTITION);
```
Would be good to move these (unrelated) cleanup change into separate commit. It makes the review more easy and also backports are much more easy to do when there's a need to backport a commit. Not anything which needs to be done now, just FIY - a suggestion for your next PRs to come.

vjuranek reviewed (commented):
Overall LGTM, just one minor suggestion for improving the test. But given my very limited knowledge of Cassandra, it would be nice to have more reviews:-)

ahmedjami commented:
@smiklosovic 

The new implementation works well for _partition key type row deletion_  when I perform a delete query with only the partition key.

However, the connector fails when I try to perform a delete with a range tombstone. Here is the error :

```
INFO  || LogicalCommitLog{commitLogPosition=CommitLogPosition(segmentId=1651752179377, position=0), synced=1048496, completed=true, log=/var/lib/cassandra/cdc_raw_directory/CommitLog-7-1651752179377.log, index=/var/lib/cassandra/cdc_raw_directory/CommitLog-7-1651752179377_cdc.idx, commitLogSegmentId=1651752179377} [io.debezium.connector.cassandra.Cassandra4CommitLogProcessor$CommitLogProcessingCallable]

2022-05-05 12:37:31,119 ERROR || Processing of event Record{source={cluster=cassandra-cluster, keyspace=keyring_prod, file=CommitLog-7-1651752179377.log, connector=cassandra, pos=514545, ts_micro=1651754179795815, version=${project.version}, snapshot=false, table=cred}, after={s={name=s, value=100,3QxeOYuvcrqVcM5SlIcBHR6RGjKRn7BLkIb_rThr, deletionTs=null, type=PARTITION}}, keySchema=Schema{io.debezium.connector.cassandra.test_prefix.keyring_prod.cred.Key:STRUCT}, valueSchema=Schema{io.debezium.connector.cassandra.test_prefix.keyring_prod.cred.Envelope:STRUCT}, op=r, ts=1651754249716} was errorneous: {} [io.debezium.connector.cassandra.QueueProcessor]

io.debezium.DebeziumException: Failed to send record Record{source={cluster=cassandra-cluster, keyspace=keyring_prod, file=CommitLog-7-1651752179377.log, connector=cassandra, pos=514545, ts_micro=1651754179795815, version=${project.version}, snapshot=false, table=cred}, after={s={name=s, value=100,3QxeOYuvcrqVcM5SlIcBHR6RGjKRn7BLkIb_rThr, deletionTs=null, type=PARTITION}}, keySchema=Schema{io.debezium.connector.cassandra.test_prefix.keyring_prod.cred.Key:STRUCT}, valueSchema=Schema{io.debezium.connector.cassandra.test_prefix.keyring_prod.cred.Envelope:STRUCT}, op=r, ts=1651754249716}

    at io.debezium.connector.cassandra.KafkaRecordEmitter.emit(KafkaRecordEmitter.java:71)

    at io.debezium.connector.cassandra.QueueProcessor.processEvent(QueueProcessor.java:114)

    at io.debezium.connector.cassandra.QueueProcessor.process(QueueProcessor.java:72)

    at io.debezium.connector.cassandra.AbstractProcessor.start(AbstractProcessor.java:63)

    at io.debezium.connector.cassandra.CassandraConnectorTaskTemplate$ProcessorGroup.lambda$start$0(CassandraConnectorTaskTemplate.java:231)

    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)

    at java.base/java.util.concurrent.FutureTask.run(Unknown Source)

    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)

    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)

    at java.base/java.lang.Thread.run(Unknown Source)

Caused by: org.apache.kafka.connect.errors.DataException: start is not a valid field name

    at org.apache.kafka.connect.data.Struct.lookupField(Struct.java:254)

    at org.apache.kafka.connect.data.Struct.put(Struct.java:202)

    at io.debezium.connector.cassandra.RowData.record(RowData.java:71)

    at io.debezium.connector.cassandra.Record.buildValue(Record.java:86)

    at io.debezium.connector.cassandra.KafkaRecordEmitter.toProducerRecord(KafkaRecordEmitter.java:81)

    at io.debezium.connector.cassandra.KafkaRecordEmitter.emit(KafkaRecordEmitter.java:62)

    ... 9 common frames omitted


```
Could you take a look please ?

smiklosovic commented:
Interesting. Yes, I will take a look. Not sure what the bug is though.

smiklosovic commented:
@ahmedjami can you play with `source.struct.version` property?

ahmedjami commented:
Here is the schema conf used by Debezium:

```
key.converter.schemas.enable=false
value.converter.schemas.enable=false

key.converter=org.apache.kafka.connect.json.JsonConverter
value.converter=org.apache.kafka.connect.json.JsonConverter
```
But, I don't set the schema registry URL (`key.converter.schema.registry.url`) so it shouldn't be used in my case.

smiklosovic commented:
ok, I need to test it more.

smiklosovic commented:
I am on it, digging the rabbit hole. The current implementation is not too much user friendly when it comes to testing internal stuff, it works nicely when we take a look at that from perspective, that works nicely, but as soon as I want to extract and test only some isolated part of the logic, it is not so straightforward. Hence I am spending some time on testability in general.

smiklosovic commented:
@ahmedjami I think I fixed it, would you mind to take a look again? I am not sure if the format I put these ranges in will be ok for you, that is why I need broader consensus on this. You will likely see it in the message sent to Kafka.

I reworked tests in such a way that it will try to construct the message as it would be sent to Kafka.

I will copy the logic done in cassandra-4 for range tombstones to cassandra-3 once we are ok with this approach.

ahmedjami commented:
Hi @smiklosovic 
Yes of course :) Will keep you updated.

smiklosovic commented:
Hey, any progress here?

ahmedjami commented:
Hey @smiklosovic 
Sorry for the delay, we have tested the fix and I don't show the error :) 
The format looks good to me but I need to discuss and share it with our dev team. It should be confirmed this week.

Thanks,

ahmedjami commented:
Hey,

@smiklosovic , the message format is OK and all tests are green :) Thanks a lot for this excellent work.

@vjuranek @jpechane PR can be merged if you don't have any other comments :)

Thanks,

vjuranek commented:
Merged, thanks @smiklosovic ! And thanks @ahmedjami for testing!

smiklosovic commented:
@vjuranek this is not all! :D I need to backport it to cassandra 3 as well, what I did was for Cassandra 4 only. If you go through the comments you will see it what I did. This is misunderstanding. 

So I ll prepare it for cassanra 3 too ok?

ahmedjami commented:
oh my bad !
I should also not have asked for it to be merged... You will reopen it or create a new ticket for that?

smiklosovic commented:
I will prepare next PR, it can go under same JIRA ticket.

vjuranek commented:
@smiklosovic yes, I understand that. Is there any need to do it in the same PR? My understanding is that your approach is fine, @ahmedjami tested it and it works, so you can go ahead and implement it aslo for Cassandra 3. Am I missing something?

smiklosovic commented:
@vjuranek that is great you get that, I was thinking that after he confirms I will just extend the same approach to Cassandra 3 and we just commit it all at once so we do not need to merge it twice. But if you do not mind me to go and create another PR, be my guest.

vjuranek commented:
@smiklosovic definitely no problem with another PR on my side. Actually I'd prefer it as going through huge PR is error prone.

vjuranek commented:
One more thing: after merging this PR, Cassandra 4 build stated to [fail](https://github.com/debezium/debezium-connector-cassandra/actions/runs/2344339482) although before merged [it was passing](https://github.com/debezium/debezium-connector-cassandra/actions/runs/2299826234). Could you please take a look @smiklosovic and eventually fix it to unbreak the build? Thanks!

vjuranek commented:
@smiklosovic I create PR #56 to fix it - fixes it for me at least locally.
